src/rmt.rs

```diff
@@ -0,0 +1,691 @@
+///
+/// Internally RMT uses pairs of pulses as part of its data structure. This implementation
+/// you need to [`set`][StackPairedSignal::set()] a two [`Pulse`]s for each index.
+///
+/// ```rust
+/// # use esp_idf_hal::rmt::StackPairedSignal;
+/// let p1 = Pulse::new(PinState::High, PulseTicks::new(10));
+/// let p2 = Pulse::new(PinState::Low, PulseTicks::new(11));
+/// let p3 = Pulse::new(PinState::High, PulseTicks::new(12));
+/// let p4 = Pulse::new(PinState::Low, PulseTicks::new(13));
+///
+/// let mut s = StackPairedSignal::new();
+/// s.set(0, &(p1, p2));
+/// s.set(1, &(p3, p4));
+/// ```
+#[derive(Clone)]
+pub struct StackPairedSignal<const N: usize>([rmt_item32_t; N]);
+
+impl<const N: usize> StackPairedSignal<N> {
+    /// Creates a new array of size `<N>`, where the number of pulses is `N * 2`.
+    pub fn new() -> Self {
+        Self(
+            [rmt_item32_t {
+                __bindgen_anon_1: rmt_item32_t__bindgen_ty_1 {
+                    // Quick way to set all 32 bits to zero, instead of using `__bindgen_anon_1`.
+                    val: 0,
+                },
+            }; N],
+        )
+    }
+
+    /// Set a pair of [`Pulse`]s at a position in the array.
+    pub fn set(&mut self, index: usize, pair: &(Pulse, Pulse)) -> Result<(), EspError> {
+        let item = self
+            .0
+            .get_mut(index)
+            .ok_or_else(|| EspError::from(ERANGE as i32).unwrap())?;
+
+        // SAFETY: We're overriding all 32 bits, so it doesn't matter what was here before.
+        let inner = unsafe { &mut item.__bindgen_anon_1.__bindgen_anon_1 };
+        inner.set_level0(pair.0.pin_state as u32);
+        inner.set_duration0(pair.0.ticks.0 as u32);
+        inner.set_level1(pair.1.pin_state as u32);
+        inner.set_duration1(pair.1.ticks.0 as u32);
+
+        Ok(())
+    }
+}
+
+impl<const N: usize> Signal for StackPairedSignal<N> {
+    fn as_slice(&self) -> &[rmt_item32_t] {
+        &self.0
+    }
+}
+
+impl<const N: usize> Default for StackPairedSignal<N> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+// TODO: impl<const N: usize> From<&[Pulse; N]> for StackSignal<{ (N + 1) / 2 }> {
+// Implementing this caused the compiler to crash!
+
+/// `Vec` heap based storage for an RMT signal.
+///
+/// Use this for when you don't know the final size of your signal data.
+///
+/// # Example
+/// ```rust
+/// let mut signal = VecSignal::new();
+/// signal.push(Pulse::new(PinState::High, PulseTicks::new(10)));
+/// signal.push(Pulse::new(PinState::Low, PulseTicks::new(9)));
+/// ```
+
+#[derive(Clone, Default)]
+#[cfg(feature = "std")]
+pub struct VecSignal {
+    items: Vec<rmt_item32_t>,
+
+    // Items contain two pulses. Track if we're adding a new pulse to the first one (true) or if
+    // we're changing the second one (false).
+    next_item_is_new: bool,
+}
+
+#[cfg(feature = "std")]
+impl VecSignal {
```

Comment:
The fact that this `Signal` is implemented using a `Vec` is an internal implementation detail, so it probably does not belong in the name of this API structure.

How about renaming this to `VariableLengthSignal` or something like that?